packages/builder/src/pages/builder/app/[application]/settings/oauth2/OAuth2ConfigModalContent.svelte

```diff
@@ -0,0 +1,144 @@
+  let hasBeenSubmitted = false
+
+  $: data = (config as Partial<OAuth2Config>) ?? {}
+
+  $: isCreation = !config
+  $: title = isCreation
+    ? "Create new OAuth2 connection"
+    : "Edit OAuth2 connection"
+
+  const requiredString = (errorMessage: string) =>
+    z.string({ required_error: errorMessage }).trim().min(1, errorMessage)
+
+  const validateConfig = (config: Partial<OAuth2Config>) => {
+    const validator = z.object({
+      name: requiredString("Name is required.").refine(
+        val =>
+          !$oauth2.configs
+            .filter(c => c.id !== config.id)
+            .map(c => c.name.toLowerCase())
+            .includes(val.toLowerCase()),
+        {
+          message: "This name is already taken.",
+        }
+      ),
+      url: requiredString("Url is required.").url(),
+      clientId: requiredString("Client ID is required."),
+      clientSecret: requiredString("Client secret is required."),
+    }) satisfies ZodType<UpsertOAuth2Config>
+
+    const validationResult = validator.safeParse(config)
+    errors = {}
+    if (!validationResult.success) {
+      errors = Object.entries(
+        validationResult.error.formErrors.fieldErrors
+      ).reduce<Record<string, string>>((acc, [field, errors]) => {
+        if (errors[0]) {
+          acc[field] = errors[0]
+        }
+        return acc
+      }, {})
+    }
+
+    return validationResult
+  }
+
+  $: saveOAuth2Config = async () => {
+    hasBeenSubmitted = true
+    const validationResult = validateConfig(data)
+    if (validationResult.error) {
+      return keepOpen
+    }
+
+    try {
+      if (isCreation) {
+        await oauth2.create(validationResult.data)
+        notifications.success("Settings created.")
+      } else {
+        await oauth2.edit(config!.id, validationResult.data)
+        notifications.success("Settings saved.")
+      }
+    } catch (e: any) {
+      notifications.error(e.message)
+      return keepOpen
+    }
+  }
+
+  $: hasBeenSubmitted && validateConfig(data)
+
+  $: isProtectedPassword = config?.clientSecret === PASSWORD_REPLACEMENT
+</script>
+
+<ModalContent onConfirm={saveOAuth2Config} size="M">
+  <Heading size="S">{title}</Heading>
+
+  <Body size="S">
+    The OAuth 2 authentication below uses the Client Credentials (machine to
+    machine) grant type.
+  </Body>
+  <Divider noGrid noMargin />
+  <Input
+    label="Name*"
+    placeholder="Type here..."
+    bind:value={data.name}
+    error={errors.name}
+  />
+  <Input
+    label="Service URL*"
+    placeholder="E.g. www.google.com"
+    bind:value={data.url}
+    error={errors.url}
+  />
+  <div class="field-info">
+    <Body size="XS" color="var(--spectrum-global-color-gray-700)">
+      The location where the flow sends the credentials. This field should be a
+      full URL.
+    </Body>
+  </div>
+  <Input
+    label="Client ID*"
+    placeholder="Type here..."
+    bind:value={data.clientId}
+    error={errors.clientId}
+  />
+  <Input
```

Comment:
One thing I believe we discussed was support for environment variable usage on the client id/secret, for production/development flagging and to make it easier to secure this information; has this been implemented yet?

Reply:
Not yet, it's in the todo list